# PLASMAR
**Introduction:**

PLASMAR (plasmid-like assembly sequence matching algorithm for resistance) is a bioinformatic tool that determines potential matches to antimicrobial resistant (AMR) carbapenemase-encoding plasmids (CEPs) from short read bacteria assemblies. It identifies a set of possible CEP matches based on shared AMR genes and plasmid replicons, and then incorporates machine learning models to estimate the probability that assemblies carry known CEPs and also if multiple assemblies carry the same CEPs.

**Quick Installation via Conda:**

PLASMAR (and all the dependencies) can be installed via Conda:
```
conda install -c conda-forge -c bioconda -c defaults plasmar
```

Run the PLASMAR.py script to see usage:
```
PLASMAR.py -h
```
If you download directly via Git, PLASMAR requires Python >=3.11, GAMMA (github.com/rastanton/GAMMA), NumPy >=1.25 (numpy.org), scikit-learn v.1.6.1 (scikit-learn.org), Entrez Direct (ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/versions/current/), and Matplotlib (matplotlib.org), all of which have to be in your $PATH.

**Initial Setup:**

Before running the first time, the plasmid database must be built using this command:
```
DB_Maker.py
```
This may take 20 minutes or more depending on the speed of your connection, but only has to be run once for the initial installation.

**Usage:**

The input for PLASMAR is a folder containing assemblies in fasta format and with a *.fasta extension (i.e., My_Assembly.fasta). It is run like so:
```
PLASMAR.py -f Path/To/My/Assembly/Folder
```
There are two optional arguments

  -m,--matches:    Only determines the potential CEP matches for each assembly and not the probabilities of multiple assemblies carrying the same CEP
  
  -a, --all:    Considers all plasmids with the same carbapenemase as potential matches (instead of just those with the same AMR genes and plasmid replicons as those found in the assembly). The models were not trained on this type of data so the probabilities may be unreliable, and are only recommended for use for samples that have few overlaps identified without this option.

Once run, PLASMAR will create a set of four output folders.
  - AMR: AMR gene matches generated by GAMMA (github.com/rastanton/GAMMA, output files in the My_Assembly_AMR.gamma format)
  - PR: The plasmid replicon matches generated by GAMMA (e.g., My_Assembly_PR.gamma)
  - PLASMAR_Matches: Tab-delimited, PLASMAR match files containing the descriptors generated and a matching probability for each potential CEP match (e.g., My_Assembly_Matches_Prob.txt)
  - Reports: This folder contains six different file types:
    1. Overlap_Comp_Probs.txt: Tab-delimited file with the descriptors and probabilities of all assembly pairs that have potential overlapping CEPs
    2. Overlap_Comp_Report.txt: A simplified, tab-delimited summary of the Overlap_Comp_Probs.txt file with the overlapping carbapenemase alleles, AMR genes, plasmid replicons, and CEP overlap probability of the assembly pairs
    3. Overlap_Comp_Report_CP-allele.txt: The same format report as the Overlap_Comp_Report.txt, but generated for each unique carbpanemase allele present in the assembliy set (i.e., Overlap_Comp_Report_blaKPC-3.txt)
    4. Plasmid_Presence.txt: A tab-delimited file with the probability that the carbapenemase genes found on each assembly are on a plasmid (vs. chromosomal)
    5. PLASMAR_Summary.txt: A tab-delimited file with the summary of PLASMAR, including the carbapenemase allele, AMR genes, plasmid replicons, and probability of the most likely CEP match for each assembly, as well as a list of other assemblies from the set that have potential overlapping CEPs (i.e., those with a probability score > 0.2)
    6. CP-allele_Heatmap.png: A .png image file of the heatmap of the probability overlaps of the assemblies carrying the same CEP for each carbapenemase allele in the assembly set (e.g., blaKPC-3_Heatmap.jpg)

There is an additional script that removes contigs generated by SAUTE (the target enriched assembler) from assemblies downloaded directly from NCBI. The SAUTE contigs are generated with the AMRFinderDB as targets, and can include additional copies of carbapenemases and other AR genes already found in the assembly. The data used to create the PLASMAR models did not include these SAUTE contigs, so it is generally recommended that these be removed before running PLASMAR, unless these contain the only copies of the carbapenemase genes. The script requires the folder with the *.fasta assembly files as input:
```
Saute_Remover_Folder.py Path/To/My/Assembly/Folder
```
Organization: CDC/NCEZID/DHQP/PRB

Contact email: ncezid_shareit@cdc.gov

Description: PLASMAR (plasmid-like assembly sequence matching algorithm for resistance) is a bioinformatic tool that determines potential matches to antimicrobial resistant (AMR) carbapenemase-encoding plasmids (CEPs) from short read bacteria assemblies. It identifies a set of possible CEP matches based on shared AMR genes and plasmid replicons, and then incorporates machine learning models to estimate the probability that assemblies carry known CEPs and also if multiple assemblies carry the same CEPs.

Languages: Python
